### ISSUE TITLE
Debug Label

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#115]
**Describe the project you are working on:**
2D action game, turn based card game.

**Describe the problem or limitation you are having in your project:**
Writing code in Godot is fast, but I actually spend the majority of my dev time debugging and making sure what I wrote actually works. Using ``print`` in all it's variants is great for the utmost basic things, but especially for any code running in _process() or _physics_process() printing to output is anything but optimal due to the nature of print: 
It consumes a lot of performance in _process() or _physics_process() severely limiting the usefulness of "live" debugging with print
it's had to keep track of a larger number of different values that change each frame

Two alternatives: 
Either print to file, has less performance issues, but requires to clutter your project with file code and is not suitable for live-debugging,
or create Label node and set_text to those values which will also require custom code that clutters your project and therefore scales really badly.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
A Debug-Label node, that serves to print all values I pass as arguments would solve all these problems:

`$DebugLabel.set_text(variable or property or signal, variable or property or signal, variable or property or signal, ...)`

would automatically format the argument already like 
argument1: value of argument1
argument2: value of argument2
argument3: value of argument3
...
according to what type the value is, the argument value is colored differently according to the users Godot editor theme. So for the default theme this would be strings are yellow, booleans are red, color is green, integers white.

It would provide immediate onscreen information, right at the location where I need it. It would have very little performance costs compared to print (presumably close to a simple Label) and provide all the Label Node comforts (scale, relative positioning, add it as child to the scene I want to debug, and therefore moving along relative with the parent scene)

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![debug_label](https://user-images.githubusercontent.com/47016402/65868844-6b2a6800-e379-11e9-990a-211ef28c170f.png)

**Describe implementation detail for your proposal (in code), if possible:**
I don't have the coding skills.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It's a debugging improvement, so I can imagine it being used a lot.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Because debugging is such an essential and big part of development time, efficient and easy use should not be an downloadable option, but already come with the engine on first installment.

**Comment by Jummit** [comment]
Another way of doing this could be a plugin that displays the log in the game, so you would still use simple print statements to debug, which is more convenient than `$DebugLabel.text += "\nFoo Bar"`.

**Comment by willnationsdev** [comment]
@Jummit I think that misses the point. The user's goal is specifically NOT to repeatedly see the same 7 parameters reprinted over and over to the console, flooding the console with information. His goal is instead to have a single place he can observe, in the world, that follows another object and which constantly displays live values for a subset of objects' data.

**Comment by golddotasksquestions** [comment]
> Another way of doing this could be a plugin that displays the log in the game
No, that would be something completely different as the print log can have all kinds of data.
As a plugin is also not part of the Godot vanilla. I would have needed this from the first time I started Godot.

**Comment by Jummit** [comment]
Both have their uses. My suggestion would be handy for single time events, and @golddotasksquestions one would be great for another way of observing variables other than the remote tree view (which is quite handy already tbh).

**Comment by willnationsdev** [comment]
Just for some insight on how something like this might be implemented:

1. Create a type that extends Label.
1. Define a set of configurable Theme values used to color-code each Variant value type.
1. Define a property that specifies the target object for it to render properties about. We'll call this the "target".
1. Export a NodePath. When assigned a non-empty value, the NodePath populates the "target" property with a Node.
1. Define a string enum PropertyInfo that, when exported, would generate a dropdown of the property names that have been defined for the target.
1. Export an Array of the above PropertyInfos so that users can specify a subset of properties for which they would like to render values.
1. During the draw notification of the DebugLabel, iterate over the target object's property list, filter it based on the exported subset of properties from the Array, switch on the Variant type of the PropertyInfo to determine the color to use from the Theme, and use `font->draw_char(...)` to render the value beside the name of the field.

**Comment by clayjohn** [comment]
In [this reddit thread](https://www.reddit.com/r/godot/comments/dbb22b/what_do_you_think_about_a_debuglable_node/) a user was able to implement this feature in GDScript in under an hour. 

I think we need a better reason for inclusion in core than "it would be useful". Especially for something that is so quick and easy to implement in GDScript. Keep in mind there needs to be a compelling reason why this should be in core and not be made into an addon. This feature could likely be made into a short script that could be found on the asset library (i'm guessing it would even be short enough to include in a tweet).

In my opinion, something like this is better suited to be part of the API proposed in https://github.com/godotengine/godot-proposals/issues/112

**Comment by willnationsdev** [comment]
@clayjohn That's fair. Curious how, with #112, you would specify that debug text for properties should be rendered at a specific location and track. Suppose you'd have to write it in `_process` or something?

**Comment by golddotasksquestions** [comment]
It's not fair at all. Only because clayohn or others can code themselves a tool within one hour to make all kinds of magic happen, does not mean everyone else who uses the engine can, nor does it mean that this is an approach that fosters faster developing, prototyping or debugging.

Addons means horrible discoverability, means unreliable documentation (because no integration in to official documentation), means not available for first time users, means eaveryone works with a different Godot (hard time to reproduce bugs, hard time to follow tutorials)
Why you want to cast features that are obviously wanted and would be incredible useful for everyone using Godot to an external plugin/addon is beyond me.
If you want to keep the codebase just the size it currently is, I would suggest to start kicking out features which are used less widely then the proposed features or are simply redundant, instead of arguing against the inclusion of features that are actually asked for and needed.

**Comment by willnationsdev** [comment]
@golddotasksquestions 

He wasn't saying we shouldn't have this feature. He is saying there is already another feature proposal that supersedes the currently proposed one, i.e. we could code a Debug Singleton of some sort that would wrap all of this functionality *and more*. He wasn't saying this wouldn't be useful. He was saying that there is already a broader and potentially more user-friendly platform by which this feature could be delivered to users.

**Comment by willnationsdev** [comment]
I would imagine something akin to this:

    extends Node
    export(NodePath) var _path_target := @".."
    func _ready() -> void:
        var target = get_node(_path_target)
        assert(target)
        # params: parent node, offset, object to target, properties on target to display
        var id = Debug.property_display_create(target, Vector2(20, 20), target, ["num", "text"])
        # ...
        Debug.property_display_set_visibility(id, false)
        Debug.property_display_delete(id)

**Comment by girng** [comment]
I don't have performance issues with `print`, and I'm running on an i7 2600, with a 144hz monitor. Sometimes printing stuff in `_process`. I also **personally** would not really use this because of visual clutter, as there is so much going in my scenes. However, I'm not saying it's a bad idea, just not for me.

The development process to create something like this (@willnationsdev's post above) is a testament at how powerful the current gdscript/node system is,

**Comment by nezvers** [comment]
> @Jummit I think that misses the point. The user's goal is specifically NOT to repeatedly see the same 7 parameters reprinted over and over to the console, flooding the console with information. His goal is instead to have a single place he can observe, in the world, that follows another object and which constantly displays live values for a subset of objects' data.

Defold has nice debugging print - if it's printing the same value, then it doesn't write new line but instead indicates call count next to the line.

**Comment by Xrayez** [comment]
For a more complete solution, I think this could be done as part of [GodotVisualDebugger](https://github.com/rocknightstudios/GodotVisualDebugger).

Also see it being promoted at GodotCon: [Visual Debugger for the Godot community](https://youtu.be/IekyN7LsHFQ). 😛

**Comment by golddotasksquestions** [comment]
@Xrayez 
I watched this GonCon talk in it's entirety already. I had no idea what to do with it afterwards or how to use it.
The point of this particular proposal right here is to discuss a solution to display variables and properties of nodes while it is running that is as straight forward and accessible and need as little explanation as ``print()``

**Comment by Xrayez** [comment]
Yeah I think it's still work-in-progress and seems like the development has staled...

I'd support this being useful for debugging networking properties at runtime, because there's no way to debug stuff visually while running several instances of the game (windows) at localhost, so I have to basically rewrite a lot of built-in editor debugging functionality.

**Comment by Xrayez** [comment]
I've implemented the most general-purpose debug label node I could think of in godot-extended-libraries/godot-next#49. The color coding is not implemented yet, see implementation description and the test project.

**Comment by avencherus** [comment]
I would think that a more robust solution that could be be more suitable as something built-in, would be a way to create an instance of the Inspector and attach it to any control node during a running game.  

It's API would allow it to point it at any resource during runtime, and you could navigate it just as you do in the editor.

You could then write a small bit of wrapper to use it like any other widget.  Instance it, put it in a popup, and have your script send it what to watch.

This though would be useful only in cases of running the application after it's built, as the live editing will suffice during development.
